language/move-vm/runtime/src/loader.rs

```diff
@@ -665,57 +665,71 @@ impl Loader {
         log_context: &impl LogContext,
     ) -> VMResult<()> {
         bytecode_verifier::verify_module(&module)?;
         self.check_natives(&module)?;
 
-        let deps = module.immediate_dependencies();
+        let (bundle_deps, cached_deps): (Vec<_>, Vec<_>) = module
+            .immediate_dependencies()
+            .into_iter()
+            .partition(|module_id| bundle_verified.contains_key(module_id));
         let loaded_imm_deps = if verify_no_missing_modules {
-            self.load_dependencies_verify_no_missing_dependencies(deps, data_store, log_context)?
+            self.load_dependencies_verify_no_missing_dependencies(
+                cached_deps,
+                data_store,
+                log_context,
+            )?
         } else {
-            self.load_dependencies_expect_no_missing_dependencies(deps, data_store, log_context)?
+            self.load_dependencies_expect_no_missing_dependencies(
+                cached_deps,
+                data_store,
+                log_context,
+            )?
         };
-        self.verify_module_dependencies(module, loaded_imm_deps)
+        let all_imm_deps = loaded_imm_deps.iter().map(|module| module.module()).chain(
+            bundle_deps
+                .iter()
+                .map(|module_id| bundle_verified.get(module_id).unwrap()),
+        );
+        dependencies::verify_module(module, all_imm_deps)
     }
 
-    fn verify_module_dependencies(
+    fn verify_module_cyclic_relations(
         &self,
         module: &CompiledModule,
-        imm_dependencies: Vec<Arc<Module>>,
+        bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
+        bundle_unverified: &BTreeSet<ModuleId>,
     ) -> VMResult<()> {
-        let imm_deps: Vec<_> = imm_dependencies
-            .iter()
-            .map(|module| module.module())
-            .collect();
-        dependencies::verify_module(module, imm_deps)
-    }
-
-    fn verify_module_cyclic_relations(&self, module: &CompiledModule) -> VMResult<()> {
         let module_cache = self.module_cache.read();
         cyclic_dependencies::verify_module(
             module,
             |module_id| {
-                module_cache
-                    .modules
+                bundle_verified
                     .get(module_id)
+                    .or_else(|| module_cache.modules.get(module_id).map(|m| m.module()))
+                    .map(|m| m.immediate_dependencies())
                     .ok_or_else(|| PartialVMError::new(StatusCode::MISSING_DEPENDENCY))
-                    .map(|m| m.module().immediate_dependencies())
             },
             |module_id| {
-                module_cache
-                    .modules
-                    .get(module_id)
-                    .ok_or_else(|| PartialVMError::new(StatusCode::MISSING_DEPENDENCY))
-                    .map(|m| m.module().immediate_friends())
+                if bundle_unverified.contains(module_id) {
+                    Ok(vec![])
```

Comment:
Would you mind adding some comments to say what each code path means?

Reply:
Explanation added